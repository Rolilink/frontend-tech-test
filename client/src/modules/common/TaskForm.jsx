import React from 'react';
import PropTypes from 'prop-types';
import { Form, FormGroup, Button, FormControl, ControlLabel } from 'react-bootstrap';


export default class TaskForm extends React.Component {
  static propTypes = {
    onSubmit: PropTypes.func.isRequired,
    onCancel: PropTypes.func.isRequired,
    onTitleChange: PropTypes.func.isRequired,
    onDescriptionChange: PropTypes.func.isRequired,
    task: PropTypes.any,
    className: PropTypes.string,
  }

  static defaultProps = {
    className: '',
    task: {
      title: '',
      description: '',
    },
  }

  onSubmit(e) {
    e.preventDefault();

    this.props.onSubmit();
  }

  isSubmitButtonEnabled() {
    const { task } = this.props;

    return task.title.length > 0 && task.description.length > 0;
  }

  render() {
    const {
      task,
      onTitleChange,
      onDescriptionChange,
      onCancel,
    } = this.props;

    return (
      <Form onSubmit={e => this.onSubmit(e)} className={`TasksForm ${this.props.className}`}>
        <FormGroup className="TasksForm-Title">
          <ControlLabel>Title:</ControlLabel>
          <FormControl
            type="text"
            onChange={e => onTitleChange(e.target.value)}
            value={task.title}
            className="TasksForm-TitleInput"
          />
        </FormGroup>
        <FormGroup className="TasksForm-Description">
          <ControlLabel>Description:</ControlLabel>
          <FormControl
            type="text"
            onChange={e => onDescriptionChange(e.target.value)}
            value={task.description}
            className="TasksForm-DescriptionInput"
          />
        </FormGroup>
        <FormGroup className="TasksForm-Actions">
          <input type="submit" className="btn btn-primary TasksForm-SubmitButton" disabled={!this.isSubmitButtonEnabled()} value="Submit" />
          <Button className="btn btn-primary TasksForm-CancelButton" onClick={() => onCancel()} bsStyle="danger">Cancel</Button>
        </FormGroup>
      </Form>
    );
  }
}
