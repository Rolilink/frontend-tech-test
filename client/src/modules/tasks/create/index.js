import connectToState from './providers/CreateTask';
import CreateTaskComponent from './components/CreateTask';

export default connectToState(CreateTaskComponent);
