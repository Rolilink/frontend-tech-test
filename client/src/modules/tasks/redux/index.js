import * as actionTypes from './actionTypes';
import * as factories from './factories';
import * as selectors from './selectors';
import * as utils from './utils';

import { fetchTasks, createTask } from './actions';

import reducerModule from './reducers';

const reducer = {
  name: 'tasks', reducer: reducerModule,
};

export {
  fetchTasks,
  createTask,
  actionTypes,
  factories,
  selectors,
  reducer,
  utils,
};
