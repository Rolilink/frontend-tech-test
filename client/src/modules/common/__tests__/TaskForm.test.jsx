import React from 'react';
import { mount } from 'enzyme';
import serializer from 'enzyme-to-json/serializer';

import { taskFactory } from '../../../__tests__/testUtils/tasks';
import TaskForm from '../TaskForm';

expect.addSnapshotSerializer(serializer);

describe('Common.TaskForm', () => {
  it('should render', () => {
    const task = taskFactory({ id: 1 });
    const wrapper = mount(<TaskForm task={task} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should have title input with value equals task.title when task is passed as props', () => {
    const task = taskFactory({ id: 1 });
    const wrapper = mount(<TaskForm task={task} />);

    expect(wrapper.find('.TasksForm-TitleInput')).toMatchSnapshot();
  });

  it('should have description input with value equals task.description when task is passed as props', () => {
    const task = taskFactory();
    const wrapper = mount(<TaskForm task={task} />);

    expect(wrapper.find('.TasksForm-DescriptionInput')).toMatchSnapshot();
  });

  it('should call onTitleChange to update title', () => {
    const event = { target: { name: 'title', value: 'updated title' } };
    const onTitleChange = jest.fn();
    const wrapper = mount(<TaskForm onTitleChange={onTitleChange} />);

    wrapper.find('.TasksForm-TitleInput').simulate('change', event);

    expect(onTitleChange).toBeCalledWith(event.target.value);
  });

  it('should call onDescriptionChange to update description', () => {
    const event = { target: { name: 'description', value: 'updated description' } };
    const onDescriptionChange = jest.fn();
    const wrapper = mount(<TaskForm onDescriptionChange={onDescriptionChange} />);

    wrapper.find('.TasksForm-DescriptionInput').simulate('change', event);

    expect(onDescriptionChange).toBeCalledWith(event.target.value);
  });

  it('should call onSubmit when submiting form', () => {
    const preventDefault = jest.fn();
    const event = { preventDefault };
    const onSubmit = jest.fn();
    const wrapper = mount(<TaskForm onSubmit={onSubmit} />);

    wrapper.find('.TasksForm').simulate('submit', event);

    expect(onSubmit).toBeCalled();
    expect(preventDefault).toBeCalled();
  });

  it('should call onCancel when clicking cancel button', () => {
    const onCancel = jest.fn();
    const wrapper = mount(<TaskForm onCancel={onCancel} />);

    wrapper.find('.TasksForm-CancelButton').simulate('click');

    expect(onCancel).toBeCalled();
  });

  it('should have a disabled submit button when task is not passed', () => {
    const wrapper = mount(<TaskForm />);

    expect(wrapper.find('.TasksForm-SubmitButton')).toMatchSnapshot();
  });

  it('should have a disabled submit button when task is on default state', () => {
    const task = { description: '', title: '' };
    const wrapper = mount(<TaskForm task={task} />);

    expect(wrapper.find('.TasksForm-SubmitButton')).toMatchSnapshot();
  });

  it('should have a enabled submit button when task is not on default state', () => {
    const task = { description: 'description', title: 'title' };
    const wrapper = mount(<TaskForm task={task} />);

    expect(wrapper.find('.TasksForm-SubmitButton')).toMatchSnapshot();
  });
});
