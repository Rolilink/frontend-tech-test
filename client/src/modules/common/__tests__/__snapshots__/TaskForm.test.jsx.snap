// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Common.TaskForm should have a disabled submit button when task is not passed 1`] = `
<input
  className="btn btn-primary TasksForm-SubmitButton"
  disabled={true}
  type="submit"
  value="Submit"
/>
`;

exports[`Common.TaskForm should have a disabled submit button when task is on default state 1`] = `
<input
  className="btn btn-primary TasksForm-SubmitButton"
  disabled={true}
  type="submit"
  value="Submit"
/>
`;

exports[`Common.TaskForm should have a enabled submit button when task is not on default state 1`] = `
<input
  className="btn btn-primary TasksForm-SubmitButton"
  disabled={false}
  type="submit"
  value="Submit"
/>
`;

exports[`Common.TaskForm should have description input with value equals task.description when task is passed as props 1`] = `
<input
  className="TasksForm-DescriptionInput form-control"
  onChange={[Function]}
  type="text"
  value="Description"
/>
`;

exports[`Common.TaskForm should have title input with value equals task.title when task is passed as props 1`] = `
<input
  className="TasksForm-TitleInput form-control"
  onChange={[Function]}
  type="text"
  value="Title"
/>
`;

exports[`Common.TaskForm should render 1`] = `
<TaskForm
  className=""
  task={
    Object {
      "description": "Description",
      "id": 1,
      "title": "Title",
    }
  }
>
  <Form
    bsClass="form"
    className="TasksForm "
    componentClass="form"
    horizontal={false}
    inline={false}
    onSubmit={[Function]}
  >
    <form
      className="TasksForm  "
      onSubmit={[Function]}
    >
      <FormGroup
        bsClass="form-group"
        className="TasksForm-Title"
      >
        <div
          className="TasksForm-Title form-group"
        >
          <ControlLabel
            bsClass="control-label"
            srOnly={false}
          >
            <label
              className="control-label"
            >
              Title:
            </label>
          </ControlLabel>
          <FormControl
            bsClass="form-control"
            className="TasksForm-TitleInput"
            componentClass="input"
            onChange={[Function]}
            type="text"
            value="Title"
          >
            <input
              className="TasksForm-TitleInput form-control"
              onChange={[Function]}
              type="text"
              value="Title"
            />
          </FormControl>
        </div>
      </FormGroup>
      <FormGroup
        bsClass="form-group"
        className="TasksForm-Description"
      >
        <div
          className="TasksForm-Description form-group"
        >
          <ControlLabel
            bsClass="control-label"
            srOnly={false}
          >
            <label
              className="control-label"
            >
              Description:
            </label>
          </ControlLabel>
          <FormControl
            bsClass="form-control"
            className="TasksForm-DescriptionInput"
            componentClass="input"
            onChange={[Function]}
            type="text"
            value="Description"
          >
            <input
              className="TasksForm-DescriptionInput form-control"
              onChange={[Function]}
              type="text"
              value="Description"
            />
          </FormControl>
        </div>
      </FormGroup>
      <FormGroup
        bsClass="form-group"
        className="TasksForm-Actions"
      >
        <div
          className="TasksForm-Actions form-group"
        >
          <input
            className="btn btn-primary TasksForm-SubmitButton"
            disabled={false}
            type="submit"
            value="Submit"
          />
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="danger"
            className="btn btn-primary TasksForm-CancelButton"
            disabled={false}
            onClick={[Function]}
          >
            <button
              className="btn btn-primary TasksForm-CancelButton btn btn-danger"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Cancel
            </button>
          </Button>
        </div>
      </FormGroup>
    </form>
  </Form>
</TaskForm>
`;
